[#63660] show attribute name in resolved subject

Kharonus opened the pull request:
# Ticket
[OP#63660](https://community.openproject.org/work_packages/63660)

# What are you trying to accomplish?
- show attribute name in resolved subject

# What approach did you choose and why?
- return array of token resolver objects from property mapper
- resolver reacts on absent context
- different reaction to nil values and not resolvable attributes

### More information

- the first commit contains the change regarding the pattern resolver and the underlying property mapper utility class. Those changes need to be reflected in the pattern input component, which will be part of a next commit.

Kharonus committed:
[#63660] show attribute name in resolved subject

- https://community.openproject.org/work_packages/63660
- return array of token resolver objects from property mapper
- resolver reacts on absent context
- different reaction to nil values and not resolvable attributes

NobodysNightmare commented on app/models/types/pattern_resolver.rb:
```diff
@@ -49,17 +51,32 @@ def resolve(work_package)
       context = token.context == :work_package ? work_package : work_package.public_send(token.context)
+      return ATTRIBUTE_PLACEHOLDER if context.nil?
+
+      attribute_resolver = @token_resolver.detect { |t| t.key == token.key }
```
:yellow_circle: Not sure how happy I am about the naming of `@token_resolver`. It seems to be simply an array of... things... Apparently an array of attribute resolvers? Couldn't we call it `@attribute_resolvers`? But it is returned by a method called `tokens_for_type`, so maybe it should be called `@tokens`.

Kharonus replied:
yep, agreed.

NobodysNightmare commented on app/models/types/patterns/token_property_mapper.rb:
```diff
@@ -31,88 +31,82 @@
-
-      alias :[] :fetch
+      # rubocop:disable Layout/LineLength
+      BASE_ATTRIBUTE_TOKENS = [
```
:yellow_circle: Adding to my naming confusion mentioned earlier: Here we have an array called `*_TOKENS` and it is filled with classes of type `AttributeResolver`...

Kharonus replied:
naming is hard ... :D Will improve with your suggestions

Kharonus replied:
I now differentiate mainly between pattern tokens and attribtue tokens.

The pattern tokens are the simple string representations persisted in a subject pattern bleuprint. The attribute tokens are object sharing the same keys, but containing logic of resolving values and localising labels.

NobodysNightmare commented on app/models/types/patterns/token.rb:
```diff
@@ -37,21 +37,15 @@ def self.build(pattern)
+          :parent
+        else
+          :work_package
+        end
```
Where is this `context` method being called? All the usages I found so far seem to call the (similar) implementation on the `AttributeMapper`. Where do we use `Token`?

Kharonus replied:
it was used before... maybe it is now obsolete? thanks for the hint, will check.

Kharonus replied:
Still used, hopefully with the rename it is now better visible.

NobodysNightmare commented on app/models/types/patterns/token_property_mapper.rb:
```diff
@@ -31,88 +31,82 @@
+        [
+          *BASE_ATTRIBUTE_TOKENS,
+          *tokenize(work_package_cfs_for(type)),
+          *tokenize(project_attributes, "project_"),
```
:green_circle: What do you think about renaming `project_attributes` to `project_cfs`? Then I would have understood earlier, that this method is (like the ones above and below) only adding custom fields to the static list of attributes, which already included _project attributes_  (not CFs) beforehand.

Kharonus replied:
sadly, project attributes is a synonym for cfs on project level.

But I agree, in here we can name them cfs consitently.

NobodysNightmare commented:
Generally looks like an improvement of the code we had before. I believe most of my questions only tough on pieces that come from working and reworking code until it fit.

:+1:

Kharonus committed:
[#63660] amend pattern input component

- component consumes new data from property mapper
- no need of building the prefix in component controller anymore
- sorting of values now happens in form component

Kharonus commented on app/components/work_packages/types/subject_configuration_component.rb:
ℹ️ preparation of hierarchical tree and sorting of fields now happen in form component.
